src/query.rs

```diff
@@ -550,26 +668,98 @@ mod tests {
-                    .into_iter()
-                    .map(|(k, v)| (k, v.into()))
-                    .collect();
-
                 registry
-                    .render_template(&template, &pretty_result)
+                    .render_template(&template, semver_violation_result)
                     .with_context(|| "Error instantiating semver query template.")
                     .expect("could not materialize template");
             }
         }
+
+        if let Some(witness) = semver_query.witness {
+            let expected_witness_text =
+                std::fs::read_to_string(format!("./test_outputs/witnesses/{query_name}.output.ron"))
+                .with_context(|| format!("Query {query_name} has a witness template, but could not load\n\
+                    `test_outputs/witnesses/{query_name}.output.ron` expected witness file, did you forget to add it?"))
+                .expect("failed to load expected witness output");
+            let expected_witnesses: BTreeMap<Cow<str>, BTreeSet<WitnessOutput>> =
+                ron::from_str(&expected_witness_text)
+                    .expect("could not parse expected witness outputs as ron format");
+
+            let actual_witnesses: BTreeMap<_, BTreeSet<_>> = transparent_actual_results
+                .iter()
+                .map(|(k, v)| {
+                    (
+                        Cow::Borrowed(k.as_str()),
+                        v.iter()
+                            .map(|values| {
+                                let Some(TransparentValue::String(filename)) = values.get("span_filename") else {
+                                    unreachable!("Missing span_filename String, this should be validated above")
+                                };
+                                let begin_line = match values.get("span_begin_line") {
+                                    Some(TransparentValue::Int64(i)) => *i as usize,
+                                    Some(TransparentValue::Uint64(n)) => *n as usize,
+                                    _ => unreachable!("Missing span_begin_line Int, this should be validated above"),
+                                };
+
+                                WitnessOutput {
+                                    filename: filename.to_string(),
+                                    begin_line,
+                                    hint: registry
+                                        .render_template(&witness.hint_template, values)
+                                        .expect("error rendering hint template"),
+                                }
+                            })
+                            .collect(),
+                    )
+                })
+                .collect();
+
+            if expected_witnesses != actual_witnesses {
+                // ideally we would use the `escape_strings` feature to pretty-print multiline
+                // strings, but this is blocked on ron releasing v0.9:
```

Comment:
Marking this TODO to make it stand out a bit more.
```suggestion
                // TODO: ideally we would use the `escape_strings` feature to pretty-print
                //       multiline strings, but this is blocked on ron releasing v0.9:
```